_Week8/Group3/simulate.do

```diff
@@ -0,0 +1,54 @@
+* first, drop any program called 'schools'
+cap prog drop schools
+
+* Now we create a program 'schools' that will store results in r()
+prog def schools, rclass
+	syntax, rho(real) // input is intracluster correlation coefficient ('ICC')
+	
+	clear // not clear all bc we need to preserve matrices in results.do
+	
+	set obs 200 // 200 clusters (schools)
+	gen school = _n // to refer to each school, which are clusters
+	gen n_students = round(75 + 10*rnormal()) // randomizing number of students (observations) per school (cluster)		obs_per_cluster
+	expand n_students // replace each school obs with "n_students" many copies
+	by school, sort: gen id = _n // assign ids to each student to differentiate them 
+	
+	// generating actual intracluster correlation in the data
+		// assuming total variance = 1
+		local sd_u = sqrt(`rho') // u represents standard dev BETWEEN clusters
+		local sd_e = sqrt(1-`rho') // e is standard dev WITHIN clusters
+		
+		// sampling u and e
+		by school (id), sort: gen u = rnormal(0, `sd_u') if _n == 1
+		by school (id): replace u = u[1] // this is measuring 'school effects', or variance BETWEEN schools. So, within a school, this is the same for all students.
```

Comment:
This (and all cluster-level characteristics) can be generated more easily by doing it _before_ using `expand`: that will also better organize your code into "cluster-level" work and "unit-level" work! This is a very clever approach though!